bikeshed/InputSource.py

```diff
@@ -157,26 +158,46 @@ def relative(self, relativePath) -> UrlInputSource:
 
 
 class FileInputSource(InputSource):
-    def __init__(self, sourceName: str, *, chroot: bool, chrootPath: Optional[str] = None):
+    def __init__(
+        self,
+        sourceName: str,
+        *,
+        chroot: bool,
+        chrootPath: Optional[str] = None,
+        tarFile: tarfile.TarFile = None,
+    ):
         self.sourceName = sourceName
         self.chrootPath = chrootPath
+        self.tarFile = tarFile
         self.type = "file"
         self.content = None
 
         if chroot and self.chrootPath is None:
             self.chrootPath = self.directory()
         if self.chrootPath is not None:
             self.sourceName = config.chrootPath(self.chrootPath, self.sourceName)
+        if tarFile:
+            assert not chroot, "with tarFile, chroot shouldn't be enabled"
+            assert not os.path.isabs(self.sourceName), "with tarFile, sourceName should be relative"
 
     def __str__(self) -> str:
         return self.sourceName
 
     def read(self) -> InputContent:
-        with open(self.sourceName, encoding="utf-8") as f:
-            return InputContent(
-                f.readlines(),
-                datetime.datetime.fromtimestamp(os.path.getmtime(self.sourceName)).date(),
-            )
+        if self.tarFile:
+            try:
+                mtime = self.tarFile.getmember(self.sourceName).get_info()["mtime"]
+            except KeyError as e:
+                raise FileNotFoundError(errno.ENOENT, "Not found inside tar file", e.args[0]) from e
+            with self.tarFile.extractfile(self.sourceName) as f:
+                # Decode the `bytes` to a `str`. (extractfile can't read as text.)
+                file_contents = f.read().decode(encoding="utf-8").splitlines(keepends=True)
```

Comment:
Rather than doing a test poke and *then* decoding the file, just put the file read inside the `try` and catch whatever error `extractfile` throws when the file doesn't exist.